c3/generator/devices.py

```diff
@@ -1005,19 +1005,18 @@ def __init__(self, **props):
         super().__init__(**props)
         self.inputs = props.pop("inputs", 1)
         self.outputs = props.pop("outputs", 1)
-        self.signal = None
-        if (
-            self.params["offset_amp"] is None
-        ):  # i.e. it was not set in the general params already
-            self.params["offset_amp"] = props.pop("offset_amp")
 
-    def process(self, instr, chan, signal: List[Dict[str, Any]]) -> Dict[str, Any]:
-        """Distort signal by adding noise."""
-        offset_amp = self.params["offset_amp"].get_value()
-        out_signal = {}
-        for k, sig in signal[0].items():
-            out_signal[k] = sig + offset_amp
-        self.signal = out_signal
+    def process(
+        self, instr: Instruction, chan: str, signal: List[Dict[str, Any]]
+    ) -> Dict[str, Any]:
+        """Distort signal by adding constant  noise."""
+        components = instr.comps
+        for comp in components[chan].values():
+            if isinstance(comp, Envelope):
+                offset_amp = comp.params["offset_amp"].get_value()
```

Comment:
This will throw an error when "offset_amp" is not in the Envelope parameters. An easy fix would be to add it in the default here: https://github.com/superplay1/c3/blob/1d685e5d8daf73640898040ee5b3ab8bcce02695/c3/signal/pulse.py#L47-L54